### PR TITLE
Template upload/download fixes, misc improvements

### PR DESCRIPTION
**Pull request by admiralmaggie**
Per our discussion, I have made the following changes:

1) Added a compare method to match two fingerprints templates in char buffers. ```compare_templates```
2) Added a new function to read system parameters. ```read_sysparam```
3) Added a soft reset function. ```soft_reset```
4) Improved debugging functions. ```_print_debug```
5) Added an example to demonstrate how to upload/download templates and store them in a file. ```examples\fingerprint_template_file_compare.py```
6) Corrected bugs in ```_get_data``` and ```_send_data``` functions.

### CODE REVIEW

File: examples/fingerprint_template_file_compare.py
Comment (jerryneedell) [inline examples/fingerprint_template_file_compare.py:25]: You will need to add the License/copyright lines here -- see the other examples
I think that will satisfy the CI failure

File: adafruit_fingerprint.py
Comment (jerryneedell) [inline adafruit_fingerprint.py]: I'm curious why this was added.

**Review by jerryneedell — CHANGES_REQUESTED** [review]
Thank for for this PR -- The clean-up and added features are very nice.
I tested this on a Raspberry Pi 4 and it worked well.
Also tested basic function on GrandCentral_M4 -- no issues
edited to add:
It looks like the functions added should also work on the other supported FingerPRint sensors.
The example has some R503 specific code (settings the LED). Would other changes be necessary for a different sensor?
Should the example be names as specifically for the R503.

I think the CI failure was due to missing License/Copyright lines in the example.
I don't have any other concerns.

File: examples/fingerprint_template_file_compare.py
Comment (jerryneedell) [inline examples/fingerprint_template_file_compare.py]: The LED control is specifically to the R503 sensor -- should this example be limited to the R503
Will it work on other supported sensors like the ZFM-20 sensor? It appears to have the same functions but i don't have access to one to test.

File: adafruit_fingerprint.py
Comment (admiralmaggie) [reply adafruit_fingerprint.py]: I'm running these on Windows PC so needed a way to get around the micropython requirements. Any suggestion on how to get around it?

File: examples/fingerprint_template_file_compare.py
Comment (admiralmaggie) [reply examples/fingerprint_template_file_compare.py:25]: I was wondering why it failed. Thanks. I will add the license.

File: examples/fingerprint_template_file_compare.py
Comment (admiralmaggie) [reply examples/fingerprint_template_file_compare.py]: Good point. Not sure about this. We can designate the example as R503.

File: adafruit_fingerprint.py
Comment (jerryneedell) [reply adafruit_fingerprint.py]: I'm not sure I understand - I thought the micropython modue was satisfied buy the Blinka interface -- at least, it is on a Raspberry Pi. 
Does the example work an a Windows system?

File: examples/fingerprint_template_file_compare.py
Comment (jerryneedell) [reply examples/fingerprint_template_file_compare.py]: Hopefully another reviewer has one and can test it. In any case, we may need an r503 version for the LED, so I'd suggest just changing the name of this example to reflect that.

File: adafruit_fingerprint.py
Comment (jerryneedell) [reply adafruit_fingerprint.py]: Ah -- I see -- the import of board is also commented out in the example so it is not using any CircuitPython specific modules.  I have no objection to the try/except -- just curious about it.

File: examples/fingerprint_template_file_compare.py
Comment (admiralmaggie) [reply examples/fingerprint_template_file_compare.py:25]: This is done.

File: adafruit_fingerprint.py
Comment (admiralmaggie) [reply adafruit_fingerprint.py]: Okay, I looked into this further, the issue was that my venv was missing Blinka. I thought setup had installed but apparently not. I removed the pseudo code. The examples work on both Linux (Ubuntu) and Windows 10.

**Comment by jerryneedell** [comment]
Now I think you need to run "black" on the files to make CI happy....
https://learn.adafruit.com/improve-your-code-with-pylint
I'm just the messenger ;-)

File: examples/fingerprint_template_file_compare.py
Comment (admiralmaggie) [reply examples/fingerprint_template_file_compare.py]: I wrapped the LED call with a try/except block. Now, Theoretically, it should simply get ignored if a board doesn't support it.

**Comment by admiralmaggie** [comment]
> Now I think you need to run "black" on the files to make CI happy....
> https://learn.adafruit.com/improve-your-code-with-pylint
> I'm just the messenger ;-)



> Now I think you need to run "black" on the files to make CI happy....
> https://learn.adafruit.com/improve-your-code-with-pylint
> I'm just the messenger ;-)

Oh boy... Let me see...

**Comment by admiralmaggie** [comment]
@jerryneedell Okay, finally got the CI finishing without errors. Needed both Pylint corrections and Black formatting.

**Review by jerryneedell — APPROVED** [review]
Tested updates on RPi 4 -- Looks good to me.

**Comment by jerryneedell** [comment]
This looks fine to me -- I'll let others have a day or two to review/comment then will merge.

**Comment by jerryneedell** [comment]
It would be very helpful is someone with a different fingerprint sensor than the R503 could test this.

**Comment by giacomoverdi** [comment]
Hi, I'm trying to use your Commit with an r503 fingerprint error but when I run fingerprint_template_compare.py the console give me this error:

----------------
Fingerprint templates:  []
Number of templates found:  0
Traceback (most recent call last):
  File "fingerprint_template_file_compare.py", line 177, in <module>
    if finger.set_sysparam(6, 2) != adafruit_fingerprint.OK:
AttributeError: 'Adafruit_Fingerprint' object has no attribute 'set_sysparam'


Can you help me?

**Comment by jerryneedell** [comment]
@giacomoverdi It sounds like you are still using the old library.
I just copied the new  adafruit_fingerprint.py to the same folder I ran fingerprint_template_file_compare.py from so it is the one that gets used.

**Comment by admiralmaggie** [comment]
Yeah, that comes across as if the library adafruit_fingerprint.py has not been updated to the latest version. Can you double check?

**Comment by giacomoverdi** [comment]
Thanks. Now it's working! It's amazing!! Is there any way to load the template0.dat to the fingerprint reader?

**Comment by admiralmaggie** [comment]
@giacomoverdi You mean permanently load it into a flash location?

**Comment by giacomoverdi** [comment]
Yes

Il giorno gio 18 feb 2021 alle 23:09 Aaron Rouhi <notifications@github.com>
ha scritto:

**Comment by admiralmaggie** [comment]
You should be able to use `finger.store_model()` to store the template into a flash location. Basically, read the template file, send it to a buffer (1 or 2 - called slots in the code) using `finger.send_fpdata()` . Then use `finger.store_model()` and reference the buffer number used previously (1 or 2) and a flash location (called `location`).

**Comment by giacomoverdi** [comment]
The template0.dat file is only a fingerprint? or fingers print are appended in template0.dat?

**Comment by admiralmaggie** [comment]
Just one fingerprint.

**Comment by jerryneedell** [comment]
@admiralmaggie I was hoping to get confirmation from someone that that the changes do not cause any issues with the non-r503 sensors. I don't see anything gin the code that should so if you are OK with it, I think we can merge this and make a new release. Are there any other changes you want to make first?

**Comment by admiralmaggie** [comment]
@jerryneedell I think I'm good for now. I did order a ZFM-20 from Adafruit but missed the UPS deadline on Friday. I should it get it early next week. I can run a few tests then.

**Comment by jerryneedell** [comment]
> @jerryneedell I think I'm good for now. I did order a ZFM-20 from Adafruit but missed the UPS deadline on Friday. I should it get it early next week. I can run a few tests then.

That would be great. We can wait until then to merge it.

**Comment by giacomoverdi** [comment]
Another question: is there a way to set a system id (or module address) for recognize each fingerprint reader?

**Comment by admiralmaggie** [comment]
@giacomoverdi Right now, the address is hardcoded to 0xFFFFFFFF. We could potentially make it user configurable but not sure the use case for it though...

**Comment by giacomoverdi** [comment]
Can you help me to make that? I need to assign each readers to each reles.

**Comment by giacomoverdi** [comment]
I have 4 readers and 4 rele. Each reader must be connected to one rele

**Comment by admiralmaggie** [comment]
@giacomoverdi The current module is not really designed to handle multiple sensors at once and I don't think it would be an easy task to make those changes. Also, It is risky to have multiple sensors on a shared UART since there is no built in collision management protocol (nothing prevents sensors from talking at once). It might be easier to connect each sensor to a low cost MCU and then "network" the MCUs with Ethernet/Wifi/RS422/RS485 etc (or use an MCU with 4 available UART port).

**Comment by admiralmaggie** [comment]
@jerryneedell I contacted Grow Technology Co. and asked for an updated datasheet since some statements didn't make sense (i.e. buffer size was incorrect). They sent me an updated one. This has some changes compare to one hosted on Adafruit. One major change is that it support up to 6 characteristic buffers (instead of 2). This means we can create a fingerprint template out of 6 scans instead of 2 that we are currently doing. This improves accuracy. Tricky change to implement so I will hold off making any changes for now.

[R503 fingerprint module user manual.pdf](https://github.com/adafruit/Adafruit_CircuitPython_Fingerprint/files/6023516/R503.fingerprint.module.user.manual.pdf)

**Comment by jerryneedell** [comment]
@admiralmaggie Wow! Thanks for the updated sheet. Do you know which version your sensor will support?  From Register 0x3A?  I guess any change would have to check that ?

**Comment by admiralmaggie** [comment]
@jerryneedell At this point, It should support all versions. I think if we decide to support the 6 buffer capability, then we need to check the version/model.

**Comment by jerryneedell** [comment]
@jepler Did you do any testing with this. If so , which sensor?

**Comment by jerryneedell** [comment]
@admiralmaggie Did you get your ZFM-20 sensor to test this with?

**Comment by admiralmaggie** [comment]
@jerryneedell I did but I haven't had a chance to run a test. It's on my list for this weekend.

**Comment by jerryneedell** [comment]
> @jerryneedell I did but I haven't had a chance to run a test. It's on my list for this weekend.

No rush -- I see it was merged - I suggest we hold off releasing the new version until you have tested it.

**Pull request merged**
